<center><img src='https://raw.githubusercontent.com/Jangrae/img/master/title3.png' width=500/></center>

# 2.문자열 자료형

- 기본 자료형이지만 이해해야 할 부분이 많아 별도로 분리해서 학습합니다.

## 2.1.문자열 사용

- Python의 문자열은 많은 편의성과 다양한 기능을 제공합니다.

**1) 문자열 표현**

- **작은 따옴표**, 또는 **큰 따옴표**를 사용해 문자열을 표현합니다.
- **len()** 함수로 문자열 길이(=포함된 문자 개수)를 확인할 수 있습니다.

In [ ]:
# 문자열 출력(Hello World)
print('Hello World')
print("Hello World")

In [ ]:
# 문자열 변수
a = 'Hello'
b = "World"

# 확인
print(a, b)

**[참고] 문자열 길이 확인**

- **len()** 함수로 문자열 길이(=포함된 문자 개수)를 확인할 수 있습니다.

In [ ]:
# 문자열 길이 확인
print(len(a))

**[참고] 여러 줄 문자열 표현**

- 문자열 앞뒤에 따옴표 세 개를 나열하거나, \\n을 사용해 여러 줄 표현이 가능합니다.
- print() 함수를 사용해야 의도한 내용이 제대로 표시됩니다.

In [ ]:
# 여러 줄 문자열 표시
s1 = "여러 줄의 문자열을 표시할 때는 \n을 사용하세요"
print(s1)

In [ ]:
s2 = """또 다른 방법은
큰 따옴표나 작은 따옴표를 세 개 사용하는 것입니다."""
print(s2)

## 2.2.문자열 연산

- 문자열 연산은 숫자 연산과 당연히 다릅니다.
- 문자열 연산에 대해 알아두면 이후에 편리하게 사용할 수 있습니다.

**1) 문자열 + 문자열 = 문자열 결합**

In [ ]:
# 문자열 변수
s1 = 'hello'
s2 = 'world'

# 문자열 더하기
hw = s1 + ' ' + s2

# 확인
print(hw)

**2) 문자열 * 정수 = 문자열 반복**

In [ ]:
# * 사용해 문자열 반복
print('*' * 17)
print('** ' + s1 + ' ' + s2 + ' **')
print('*' * 17)

In [1]:
# 문자열에 0을 곱하면 문자열이 제거됨
s = 'hello'
print(s * 1)
print(s * 0)
print(s * 2)

hello

hellohello


## 2.3.문자열 내 값 추가

- 문자열 내에 다른 값을 추가하고 싶을 때 **f-String**을 사용합니다.
- 문자열 앞에 **소문자 f**를 붙이고, 문자열 안 **중괄호 {}** 안에 변수 이름을 대입합니다.

In [3]:
# 다양한 변수
name = '홍길동'
age = 25
score = 2345.6789

# 확인
print(f'이름: {name}')
print(f'나이: {age}')
print(f'점수: {score}')
print(f'{age}살의 {name}의 점수는 {int(score)}입니다.')

이름: 홍길동
나이: 25
점수: 2345.6789
25살의 홍길동의 점수는 2345입니다.


**[참고] 형식 지정**

- 출력되는 형태를 나름으로 조절할 수 있습니다. (문자열:s, 정수:d, 실수:f)

In [4]:
# 확인
print(f'{age}살의 {name}의 점수는 {score:,.2f}입니다.')

25살의 홍길동의 점수는 2,345.68입니다.


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20.png' width=120 align="left"/>


[문1] 우선 다음 코드를 실행하여 점수 합계와 평균을 구하세요.

In [7]:
# 변수에 값 대입
name = '홍길동'
age = 25
score1 = 153
score2 = 204
score3 = 259
score4 = 221
score5 = 267

# 합계
tot_score = score1 + score2 + score3 + score4 + score5
print(tot_score)

# 평균
avg_score = tot_score / 5
print(avg_score)

# 연령대
age_group = age // 10 * 10
print(age_group)

1104
220.8
20


[문2]  f-String을 사용해 위 변수들 값이 다음과 같은 형태로 출력되도록 코드를 작성하세요.

~~~
이름: 홍길동
연령대: 20대
점수: 153, 204, 259, 221, 267
합계: 1,104 / 평균: 220.80
~~~


In [ ]:
# 확인





<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20e.png' width=45 align="left"/>

## 2.4.문자열 관련 주요 메서드

- 변수(또는 개체) 뒤에 '.함수' 형태로 사용하는 함수를 **메서드(Method)** 라고 부릅니다.
- 아래 언급된 메서드 외에도 문자열 관련 메서드가 상당히 많으니 관련 자료를 참고하기 바랍니다.
- 문자열을 변환하는 메서드를 사용해도 그 결과가 변수에 반영되지는 않습니다.

In [8]:
# 대문자로 변환
s = 'I have a Dream.'
print(s.upper())
print(s)

I HAVE A DREAM.
I have a Dream.


- 변환 결과를 변수에 반영하려면 메서드가 반환하는 값을 변수에 대입해야 합니다.

In [9]:
# 대문자로 변환
s = 'I have a Dream.'
print(s.upper())

# 변수에 반영
s = s.upper()
print(s)

I HAVE A DREAM.
I HAVE A DREAM.


**1) 대/소문자 변환**

- **capitalize()** 메서드: 문자열 첫 문자만 대문자, 나머지 모두 소문자로 변환합니다.
- **title()** 메서드: 단어 첫 문자만 대문자, 나머지는 모두 소문자로 변환합니다.
- **upper()** 메서드: 모든 문자를 대문자로 변환합니다.
- **lower()** 메서드: 모든 문자를 소문자로 변환합니다.

In [10]:
# 대/소문자 변환
s = 'I have a Dream.'
print(s.capitalize())
print(s.title())
print(s.upper())
print(s.lower())

I have a dream.
I Have A Dream.
I HAVE A DREAM.
i have a dream.


**2) 일부 변경 및 제거**

- **replace(x, y)** 메서드: 문자열에 포함된 일부 문자열 x를 문자열 y로 변경합니다.
- **strip(x)** 메서드: 문자열에서 양쪽의 공백 또는 특수 문자를 제거합니다.

In [11]:
# 문자열 변경
s = 'nowhere'
print(s.replace('now', 'now '))

now here


In [16]:
# 앞 뒤 빈 문자열, 특수 문자 제거
print('  world '.strip())
print('#world#$$#'.strip())
print('#world#$$#'.strip('$#'))
print('#world#$$#'.strip('#$'))
print('#world#$$#'.strip('#'))

world
#world#$$#
world
world
world#$$


- replace() 메서드를 사용하는 경우가 많으니 꼭 기억합니다.

In [17]:
# 문자열 변수
num = '123,456'

# 콤마 제거 후 정수로 변환
num = num.replace(',', '')
num = int(num)

# 연산
print(num * 2)

246912


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20.png' width=120 align="left"/>


[문1] 다음 코드의 실행 결과가 '파이썬을 열심히 공부할께요!'가 되게 수정해 보세요.



In [18]:
# 문자열 변수
my_str = '파이썬을 대충 공부할께요!'
my_str = my_str.replace("대충", "열심히")
# 문자열 변환
print(my_str)

파이썬을 열심히 공부할께요!


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20e.png' width=45 align="left"/>

**3) 분리와 연결**

- **split(x)** 메서드: x 문자를 구분자로 하여 문자열을 분리하여 리스트로 변환합니다.(공백이 기본 구분자)
- **x.join(s)** 메서드: x를 구분자로 하여 요소들을 연결한 문자열로 변환합니다.

In [19]:
# 문자열 변수
s = 'I have a Dream!'

# 문자열 분리
print(s.split())

['I', 'have', 'a', 'Dream!']


In [27]:
# 문자열 변수
s = '1/2/3/4/5/6/7/8/9'

# 문자열 분리
s_new = s.split('/')
s_new1 = '/'.join(s_new)
print(s_new1)

# 확인
print(s_new)

1/2/3/4/5/6/7/8/9
['1', '2', '3', '4', '5', '6', '7', '8', '9']


In [38]:
# 문자열 변수
s = '2020-09-14'

# 문자열 분리
year = s.split('-')[0]
month = s.split('-')[1]
day = s.split('-')[2]

# 확인
print(f'오늘은 {year}년 {month}월 {day}일 입니다.')

오늘은 2020년 09월 14일 입니다.


- join() 메서드로 구분자를 두어 연결된 문자열을 만듭니다.

In [22]:
# 요소를 연결해서 문자열로
a = ['홍길동', '한사랑', '강우동', '박여인']
s = '/'.join(a)

# 확인
print(s)

홍길동/한사랑/강우동/박여인


In [23]:
# 문자열을 구분자로 나누어 리스트로
b = s.split('/')

# 확인
print(b)

['홍길동', '한사랑', '강우동', '박여인']


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20.png' width=120 align="left"/>


[문1] 다음 주어진 문자열을 적절한 구분자를 기준으로 분리하세요.



In [29]:
# 문자열 변수
my_str = '15, 23, 52, 41, 25, 60, 47, 28, 49, 60'

# 문자열 분리
my_str = my_str.split(', ')
print(my_str)

['15', '23', '52', '41', '25', '60', '47', '28', '49', '60']


[문2] 다음 주어진 리스트의 요소를 yyyy-mm-dd 형태로 연결해 dt 변수에 담고 결과를  확인하세요.



In [31]:
# 리스트 만들기
my_list = ['2020', '09', '14']

# 문자열 연결
my_new_list = '-'.join(my_list)

# 확인
print(my_new_list)

2020-09-14


[문3] 위에서 배운 문자열 관련 메서드를 사용해 다음 주어진 리스트의 요소를 모두 대문자로 바꾸세요.


In [37]:
# 리스트 만들기
a = ['h', 'a', 'p', 'p', 'y']

# 대문자로 변경
a_new = ' '.join(a)
a_new = a_new.upper()
a_new = a_new.split(' ')

# 확인
print(a_new)

['H', 'A', 'P', 'P', 'Y']


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_20e.png' width=45 align="left"/>

<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_30.png' width=120 align="left"/>

[문1] 변수 a에 {}를 넣은 문자열 '이름: {}, 나이: {}'를 대입하고 .format()으로 적당한 값을 넣어 출력하세요.

In [39]:
a = '이름: {}, 나이: {}'
print(a.format("홍길동", 20))

이름: 홍길동, 나이: 20


[문2] 밖에서 선언한 변수 값을 문자열 안에 넣어서 출력해 봅니다.

- 변수 name에는 이름, age에는 나이 값을 대입
- 변수 a에는 f'문자열{변수}...' 형태를 사용해 위와 동일한 결과가 나오도록 출력

In [41]:
name = "김현준"
age = 25
a = f"이름: {name}, 나이: {age}"
print(a)

이름: 김현준, 나이: 25


[문3] 변수 b에 {}를 넣은 문자열 '{}은(는) {}원입니다.'를 대입하고 .format()으로 적당한 값을 넣어 출력하세요.

In [42]:
b = '{}은 {}원입니다'

print(b.format('icecream', 1000))


icecream은 1000원입니다


[문4] 밖에서 선언한 변수 값을 문자열 안에 넣어서 출력하세요.

- 변수 product에는 제품명, price에는 가격을 대입
- 변수 b에는 f'문자열{변수}...' 형태를 사용해 위와 동일한 결과가 나오도록 출력

In [44]:
product = 'icecream'
price = 1000
b = f'{product}은 {price}원입니다.'
print(b)

icecream은 1000원입니다.


[문5] 아래와 같은 결과가 나오도록 변수 c에 f'''문자열{변수}...''' 형태를 사용해 문자열을 대입하여 출력하세요.

- 단, 변수 product에 '아메리카노', count에 3, price에 4000을 대입합니다.


        주문할게요.
        아메리카노 3잔 주세요.
        한 잔이 4000원 맞죠?


In [45]:
product = '아메리카노'
count = 3
price = 4000

c = f'주문할게요.\n{product} {count}잔 주세요.\n한 잔이 {price}원 맞죠?'
print(c)


주문할게요.
아메리카노 3잔 주세요.
한 잔이 4000원 맞죠?


[문6] 변수 a에 '나는 파이썬을 배우고 있다.'를 대입하고 공백을 기준으로 분할하여 리스트 형태로 출력하세요.

- 힌트: 문자열.split(분할의 기준이 될 값)

In [46]:
a = '나는 파이썬을 배우고 있다.'
print(a.split(' '))


['나는', '파이썬을', '배우고', '있다.']


<img src='https://raw.githubusercontent.com/jangrae/img/master/practice_30e.png' width=45 align="left"/>